### DISCUSSION TITLE
Access MkDocsConfig from custom fence callback?

### DISCUSSION THREAD

**Discussion by alexvoss** [facelessuser/pymdown-extensions#2382] [Q&A] [answered]
I am writing a bunch of custom fence implementations and some of them would need to be configured in `mkdocs.yml`. I don't see the `MkDocsConfig` as part of the arguments to the formatter and validator callbacks. Is there a way to access the current configuration? If not, would it be possible to add it to the callback argument list?

**Comment by facelessuser** [comment, marked as answer]
@alexvoss Python Markdown is generally not aware of the specifics of the user of it. Python Markdown is not dependent on MkDocs, it is the other way around.

I'm not saying it is impossible to make it aware, but do you have examples of how you wish to leverage custom fences? It may help so that we can better understand your goal and explore the best options.

**Reply by alexvoss** [reply]
Thanks, I guess I should have guessed that. What I am doing right now is to store the config as part if my plugin class and retrieve it from there. When I wrote my question, I was not sufficiently aware that a plugin in MkDocs is a class and not just a bunch of functions. Perhaps that is the way it should be done.

My use case is that I want people to be able to edit structured data using the Markdown and YAML support in IDEs and then compile the result into Markdown or HTML. 

For example, I have a schema for specifying rights (think human rights but also more mundane ones) in YAML, which I want to compile into Markdown for further processing. 

Another custom fence formatter I am working on is one that takes the YAML, parses it and puts it into a Jinja2 environment, to be turned into output according to a user-supplied template. Yet another will turn YAML with a regular structure into tables, more complex ones than the Markdown table syntax supports.

Most of these formatters will have configurable options, so I was looking for a way to access those. 

I love the custom fences functionality. The fact that there is at least some editor support for writing those YAML snippets means I don't have to produce editor support for the specific use  cases I want to support. That is a huge plus for me.

Some more feedback: it did take a while for me to figure out that I have to write a validator to be able to have options and arguments after the language specification.

**Reply by facelessuser** [reply]
> My use case is that I want people to be able to edit structured data using the Markdown and YAML support in IDEs and then compile the result into Markdown or HTML.

I'm not quite sure where the IDE comes into this or how is linked into MkDocs.

My responses below will be based on my limited understanding of your use cases. It is possible I could provide more in-depth suggestions, but I'd really have to understand your specific use case in more depth to do so.

---

At least for turning structured data into tables, I think fenced code could probably work, I'm not quite sure what kind of configuration is needed from MkDocs, but one approach could work as follows. You need a MkDocs portion, so if you can intercept the Markdown object, you can set an attribute to the Markdown object that contains the config. Then you can access the attribute on the Markdown object that's passed into the custom fence.

I'm not sure if custom fences is the way I would go for everything you listed, but I also don't understand the various use cases, but if you are able to coerce it to do what you need, then it is probably fine.

---

Another approach would be to simply utilize the custom fence to preserve the data and then post process the content in a MkDocs plugin. This is kind of how things like Arithmatex work, only that it is post-processed in the browser via JavaScript instead of being post-processed in a MkDocs plugin.

**Reply by alexvoss** [reply]
Thanks, I think the solution to store the config in the plugin class works for me. 

The use cases are under development and perhaps it is a bit difficult to explain them in the abstract. If you are interested in the use of custom fences then I can ping you when they come out of stealth mode. 

Thanks for the help. I now know that I did not miss anything obvious.
